In [1]:
from sklearn.datasets import load_diabetes
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
from sklearn.model_selection import train_test_split

In [2]:
X, y = load_diabetes(return_X_y=True)

In [3]:
X.ndim

2

In [4]:
print(X.shape), print(y.shape)

(442, 10)
(442,)


(None, None)

In [5]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=2)

In [6]:
basic_lreg = LinearRegression()

In [7]:
basic_lreg.fit(X_train, y_train)

LinearRegression()

In [8]:
print(basic_lreg.coef_)
print(basic_lreg.intercept_)

[  -9.15865318 -205.45432163  516.69374454  340.61999905 -895.5520019
  561.22067904  153.89310954  126.73139688  861.12700152   52.42112238]
151.88331005254167


In [9]:
basic_y_pred = basic_lreg.predict(X_test)
r2_score(y_test, basic_y_pred)

0.4399338661568969

In [31]:
class BatchGDRegressor:

    def __init__(self, learning_rate = 0.01, epochs = 100):

        self.coef_ = None
        self.intercept_ = None
        self.lr = learning_rate
        self.epochs = epochs

    def fit (self, X_train, y_train):

        #initilize the coff and intercept
        self.coef_ = np.zeros(X_train.shape[1])
        self.intercept_ = 0

        

        for i in range(self.epochs):
            #Update the parameters
            y_hat = self.intercept_ + np.dot(X_train, self.coef_)
            intercept_der = -2 * np.mean(y_train - y_hat)
            self.intercept_ = self.intercept_ - self.lr * intercept_der

            cof_der = -2 * np.dot((y_train - y_hat), X_train)/X_train.shape[0]
            self.coef_ = self.coef_ - self.lr  * cof_der
        print(self.intercept_, self.coef_)                                                             

    def predect(self, X_test):

            return self.intercept_ + np.dot(X_test, self.coef_)
        





        

In [37]:
bgdr = BatchGDRegressor(epochs=100, learning_rate=0.01)

In [38]:
bgdr.fit(X_train,  y_train)

130.52167869517368 [ 1.65186186  0.19917482  4.07446428  3.32850514  1.46553338  1.00928161
 -2.49505736  2.61837335  4.25883531  2.52960352]


In [39]:
y_pred = bgdr.predect(X_test)

In [40]:
r2_score(y_test, y_pred )

-0.13119137039796125

Since r2score is very low , so we will tune it by changing learning rate and epochs(number of iterration)

In [41]:
bgdr = BatchGDRegressor(epochs=100, learning_rate=0.1)
bgdr.fit(X_train,  y_train)
y_pred = bgdr.predect(X_test)
print(r2_score(y_test, y_pred ))

150.74619609373707 [ 15.12707339   2.68353022  40.80700513  31.70704106  13.95950214
  10.60022843 -25.7419248   26.81308147  40.36650938  24.34424513]
0.10177953728118139


In [42]:
bgdr = BatchGDRegressor(epochs=1000, learning_rate=0.1)
bgdr.fit(X_train,  y_train)
y_pred = bgdr.predect(X_test)
print(r2_score(y_test, y_pred ))

151.9397754799712 [  62.01595071  -24.69520251  262.06008894  191.96192535   39.42148032
   10.06876863 -143.63437615  124.00613934  244.11015664  119.10986714]
0.3973309089620952


In [43]:
bgdr = BatchGDRegressor(epochs=1000, learning_rate=0.01)
bgdr.fit(X_train,  y_train)
y_pred = bgdr.predect(X_test)
print(r2_score(y_test, y_pred ))

150.7459583895869 [ 15.11346476   2.67488285  40.78586564  31.6889657   13.94336728
  10.58428515 -25.72481547  26.79062733  40.34316231  24.32553932]
0.10171086687408593


In [48]:
#This Result, we take into consideration because R2Score is Higher then basic linear regression
bgdr = BatchGDRegressor(epochs=1000, learning_rate=0.5)
bgdr.fit(X_train,  y_train)
y_pred = bgdr.predect(X_test)
print(r2_score(y_test, y_pred ))

152.01346098793306 [  14.42298327 -173.91981704  491.40676142  323.96261836  -39.01589446
 -116.05536736 -194.64650666  102.97600026  451.54201731   97.58244503]
0.45347905609096006


In [47]:
bgdr = BatchGDRegressor(epochs=1000, learning_rate=0.8)
bgdr.fit(X_train,  y_train)
y_pred = bgdr.predect(X_test)
print(r2_score(y_test, y_pred ))

151.95717195272528 [ 9.83971694e-02 -1.99254330e+02  5.22298768e+02  3.37354342e+02
 -5.39211148e+01 -1.29453145e+02 -1.92129399e+02  9.12754551e+01
  4.96807458e+02  7.54613047e+01]
0.44931366052769695


### Conclusion:

- R2Score with basic Linear regression is 39%
- And we had achived to 45% with epochs 1000 and learning rate 0.5